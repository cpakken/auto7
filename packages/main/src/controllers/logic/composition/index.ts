export * from "./connections-map"
export * from "./paths"
export * from "./blocks"
export * from "./composition"
