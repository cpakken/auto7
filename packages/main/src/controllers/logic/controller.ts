import { getEnv, getRoot, IAnyStateTreeNode, Instance, SnapshotIn, SnapshotOut, types } from "mobx-state-tree"
import { LogicComposedStore, LogicNativeStore } from "."

export interface ILogicControllerSnapshotIn extends SnapshotIn<typeof LogicController> {}
export interface ILogicControllerSnapshot extends SnapshotOut<typeof LogicController> {}
export interface ILogicController extends Instance<typeof LogicController> {}
export const LogicController = types.model({
  composed: types.optional(LogicComposedStore, {}),
  native: types.optional(LogicNativeStore, {}),
})

export function getLogicController(instance: IAnyStateTreeNode): ILogicController {
  return getEnv(instance).logicController || getRoot(instance)
}
