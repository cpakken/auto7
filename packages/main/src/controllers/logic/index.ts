export * from "./info"
export * from "./composition"
export * from "./stores"
export * from "./controller"
