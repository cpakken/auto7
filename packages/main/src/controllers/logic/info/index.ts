export * from "./interface"
export * from "./info"