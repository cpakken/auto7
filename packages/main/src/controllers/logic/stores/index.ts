export * from "./native-store"
export * from "./composed-store"
export * from "./LogicNative"
export * from "./LogicComposed"
