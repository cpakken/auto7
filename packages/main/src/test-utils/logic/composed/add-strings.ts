import { LogicComposedSnapshotIn } from "@main/controllers/logic"

export const ADDSTRING: LogicComposedSnapshotIn = {
  _id: "ADDSTRING",
  composition: {
    blocks: [{ _id: "A", logicKey: "_std_ADDSTRING" }],
    paths: [
      { _id: "pA", from: { node: "in1" }, to: { block: "A", node: "in1" } },
      { _id: "pB", from: { node: "in2" }, to: { block: "A", node: "in2" } },
      { _id: "pC", from: { block: "A", node: "out" }, to: { node: "out1" } },
    ],
  },
  info: {
    label: "ADDSTRING COMPOSED",
    inputs: [
      { _id: "in1", label: "in1", typeKey: "_std_string" },
      { _id: "in2", label: "in2", typeKey: "_std_number" },
    ],
    outputs: [{ _id: "out1", label: "out1", typeKey: "_std_string" }],
  },
}

export const ADDSTRING_NESTED: LogicComposedSnapshotIn = {
  _id: "ADDSTRING_NESTED",
  composition: {
    blocks: [{ _id: "A", logicKey: "ADDSTRING" }],
    paths: [
      { _id: "pA", from: { node: "in1" }, to: { block: "A", node: "in1" } },
      { _id: "pB", from: { node: "in2" }, to: { block: "A", node: "in2" } },
      { _id: "pC", from: { block: "A", node: "out1" }, to: { node: "out1" } },
    ],
  },
  info: {
    label: "ADDSTRING NESTED",
    inputs: [
      { _id: "in1", label: "in1", typeKey: "_std_string" },
      { _id: "in2", label: "in2", typeKey: "_std_number" },
    ],
    outputs: [{ _id: "out1", label: "out1", typeKey: "_std_string" }],
  },
}

export const ADDSTRING_COMPOUND: LogicComposedSnapshotIn = {
  _id: "ADDSTRING_COMPOUND",
  composition: {
    blocks: [
      { _id: "A", logicKey: "ADDSTRING", xy: [0, 0] },
      { _id: "B", logicKey: "ADDSTRING_NESTED", xy: [0, 0] },
      { _id: "C", logicKey: "_std_ADDSTRING", xy: [0, 0] },
    ],
    paths: [
      { _id: "p0", from: { node: "in1" }, to: { block: "A", node: "in1" } },
      { _id: "p1", from: { node: "in2" }, to: { block: "A", node: "in2" } },
      { _id: "p2", from: { node: "in2" }, to: { block: "B", node: "in1" } },
      { _id: "p3", from: { node: "in3" }, to: { block: "B", node: "in2" } },
      { _id: "p4", from: { block: "A", node: "out1" }, to: { block: "C", node: "in1" } },
      { _id: "p5", from: { block: "B", node: "out1" }, to: { block: "C", node: "in2" } },
      { _id: "p6", from: { block: "C", node: "out" }, to: { node: "out1" } },
    ],
  },
  info: {
    label: "ADDSTRING COMPOUND",
    inputs: [
      { _id: "in1", label: "in1", typeKey: "_std_string" },
      { _id: "in2", label: "in2", typeKey: "_std_string" },
      { _id: "in3", label: "in3", typeKey: "_std_string" },
    ],
    outputs: [{ _id: "out1", label: "out1", typeKey: "_std_string" }],
  },
}
