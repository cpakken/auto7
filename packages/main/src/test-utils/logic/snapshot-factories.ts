import { LogicNativeSnapshotIn, LogicComposedSnapshotIn, ILogicComposedStoreSnapshotIn } from "@main/controllers/logic"

export function createLogicComposedStoreSnapshot(
  snapshots: LogicComposedSnapshotIn | LogicComposedSnapshotIn[]
): ILogicComposedStoreSnapshotIn {
  return (Array.isArray(snapshots) ? snapshots : [snapshots]).reduce(
    (a, snapshot) => {
      const { _id, info, composition } = snapshot
      a.info.push({ _id, ...info })
      a.composition.push({ _id, ...composition })
      return a
    },
    { info: [] as any[], composition: [] as any[] }
  )
}

export function createLogicNativeStoreSnapshot(
  snapshots: LogicNativeSnapshotIn | LogicNativeSnapshotIn[]
): ILogicComposedStoreSnapshotIn {
  return (Array.isArray(snapshots) ? snapshots : [snapshots]).reduce(
    (a, snapshot) => {
      const { _id, info } = snapshot
      a.info.push({ _id, ...info })
      return a
    },
    { info: [] as any[] }
  )
}

export function addIdsToSnapshot(snapshot: LogicComposedSnapshotIn) {
  const { _id, info, composition } = snapshot
  return {
    _id,
    info: { _id, ...info },
    composition: { _id, ...composition },
  }
}
