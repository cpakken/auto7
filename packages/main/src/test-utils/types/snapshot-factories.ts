import { ITypeComposedStoreSnapshotIn, TypeNativeSnapshotIn } from "@main/controllers/types"

// export function createLogicComposedStoreSnapshot(
//   snapshots: LogicComposedSnapshotIn | LogicComposedSnapshotIn[]
// ): ILogicComposedStoreSnapshotIn {
//   return (Array.isArray(snapshots) ? snapshots : [snapshots]).reduce(
//     (a, snapshot) => {
//       const { _id, info, composition } = snapshot
//       a.info.push({ _id, ...info })
//       a.composition.push({ _id, ...composition })
//       return a
//     },
//     { info: [] as any[], composition: [] as any[] }
//   )
// }

export function createTypeNativeStoreSnapshot(
  snapshots: TypeNativeSnapshotIn | TypeNativeSnapshotIn[]
): ITypeComposedStoreSnapshotIn {
  return (Array.isArray(snapshots) ? snapshots : [snapshots]).reduce(
    (a, snapshot) => {
      const { _id, info } = snapshot
      a.info.push({ _id, ...info })
      return a
    },
    { info: [] as any[] }
  )
}

// export function addIdsToSnapshot(snapshot: LogicComposedSnapshotIn) {
//   const { _id, info, composition } = snapshot
//   return {
//     _id,
//     info: { _id, ...info },
//     composition: { _id, ...composition },
//   }
// }
