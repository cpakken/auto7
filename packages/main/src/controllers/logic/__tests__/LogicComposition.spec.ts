import { mapIter } from "@utils/iterable-fns"
import { toJS } from "mobx"
import { ILogicComposition, LogicComposition } from ".."
import { getSnapshot } from "mobx-state-tree"
import { addIdsToSnapshot } from "@main/test-utils/logic/snapshot-factories"
import { ADDSTRING_COMPOUND } from "@main/test-utils/logic/composed"

// function maptoJS(obj) {
//   return mapValues(obj, (x) => toJS(x))
// }

describe("Logic Composition", () => {
  let composition: ILogicComposition
  beforeEach(() => {
    const snapshot = addIdsToSnapshot(ADDSTRING_COMPOUND)
    composition = LogicComposition.create(snapshot.composition)
  })

  test("create model", () => {
    expect(getSnapshot(composition)).toMatchSnapshot()
  })

  test("verify path initialization in blocks", () => {
    const { inputs: inMap, outputs: outMap } = composition
    const { inputs, outputs } = composition.blocks.get("A")!

    expect(toJS(inMap)).toMatchSnapshot()
    expect(toJS(outMap)).toMatchSnapshot()

    expect(toJS(inputs)).toMatchSnapshot()
    expect(toJS(outputs)).toMatchSnapshot()
  })

  test("verify links when path delete", () => {
    const { inputs, outputs } = composition.blocks.get("A")!
    const { inputs: _inMap, outputs: _outMap } = composition

    mapIter(composition.paths.store.values(), (path) => {
      composition.paths.remove(path)
    })

    expect(getSnapshot(composition)).toMatchSnapshot()

    expect(toJS(_inMap)).toMatchSnapshot()
    expect(toJS(_outMap)).toMatchSnapshot()

    expect(toJS(inputs)).toMatchSnapshot()
    expect(toJS(outputs)).toMatchSnapshot()
  })

  test("verify links when block delete", () => {
    const { inputs, outputs } = composition

    composition.blocks.remove(composition.blocks.get("A")!)

    expect(getSnapshot(composition)).toMatchSnapshot()
    expect(toJS(inputs)).toMatchSnapshot()
    expect(toJS(outputs)).toMatchSnapshot()
  })
})
