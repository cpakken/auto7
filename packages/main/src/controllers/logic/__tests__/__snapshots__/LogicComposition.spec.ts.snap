// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logic Composition create model 1`] = `
Object {
  "_id": "ADDSTRING_COMPOUND",
  "blocks": Array [
    Object {
      "_id": "A",
      "logicKey": "ADDSTRING",
      "xy": Array [
        0,
        0,
      ],
    },
    Object {
      "_id": "B",
      "logicKey": "ADDSTRING_NESTED",
      "xy": Array [
        0,
        0,
      ],
    },
    Object {
      "_id": "C",
      "logicKey": "_std_ADDSTRING",
      "xy": Array [
        0,
        0,
      ],
    },
  ],
  "paths": Array [
    Object {
      "_id": "p0",
      "from": Object {
        "node": "in1",
      },
      "to": Object {
        "block": "A",
        "node": "in1",
      },
    },
    Object {
      "_id": "p1",
      "from": Object {
        "node": "in2",
      },
      "to": Object {
        "block": "A",
        "node": "in2",
      },
    },
    Object {
      "_id": "p2",
      "from": Object {
        "node": "in2",
      },
      "to": Object {
        "block": "B",
        "node": "in1",
      },
    },
    Object {
      "_id": "p3",
      "from": Object {
        "node": "in3",
      },
      "to": Object {
        "block": "B",
        "node": "in2",
      },
    },
    Object {
      "_id": "p4",
      "from": Object {
        "block": "A",
        "node": "out1",
      },
      "to": Object {
        "block": "C",
        "node": "in1",
      },
    },
    Object {
      "_id": "p5",
      "from": Object {
        "block": "B",
        "node": "out1",
      },
      "to": Object {
        "block": "C",
        "node": "in2",
      },
    },
    Object {
      "_id": "p6",
      "from": Object {
        "block": "C",
        "node": "out",
      },
      "to": Object {
        "node": "out1",
      },
    },
  ],
}
`;

exports[`Logic Composition verify links when block delete 1`] = `
Object {
  "_id": "ADDSTRING_COMPOUND",
  "blocks": Array [
    Object {
      "_id": "B",
      "logicKey": "ADDSTRING_NESTED",
      "xy": Array [
        0,
        0,
      ],
    },
    Object {
      "_id": "C",
      "logicKey": "_std_ADDSTRING",
      "xy": Array [
        0,
        0,
      ],
    },
  ],
  "paths": Array [
    Object {
      "_id": "p2",
      "from": Object {
        "node": "in2",
      },
      "to": Object {
        "block": "B",
        "node": "in1",
      },
    },
    Object {
      "_id": "p3",
      "from": Object {
        "node": "in3",
      },
      "to": Object {
        "block": "B",
        "node": "in2",
      },
    },
    Object {
      "_id": "p5",
      "from": Object {
        "block": "B",
        "node": "out1",
      },
      "to": Object {
        "block": "C",
        "node": "in2",
      },
    },
    Object {
      "_id": "p6",
      "from": Object {
        "block": "C",
        "node": "out",
      },
      "to": Object {
        "node": "out1",
      },
    },
  ],
}
`;

exports[`Logic Composition verify links when block delete 2`] = `
Map {
  "in1" => Array [],
  "in2" => Array [
    Object {
      "block": Object {
        "_id": "B",
        "logicKey": "ADDSTRING_NESTED",
        "xy": Array [
          0,
          0,
        ],
      },
      "node": "in1",
      "path": Object {
        "_id": "p2",
        "from": Object {
          "node": "in2",
        },
        "to": Object {
          "block": "B",
          "node": "in1",
        },
      },
    },
  ],
  "in3" => Array [
    Object {
      "block": Object {
        "_id": "B",
        "logicKey": "ADDSTRING_NESTED",
        "xy": Array [
          0,
          0,
        ],
      },
      "node": "in2",
      "path": Object {
        "_id": "p3",
        "from": Object {
          "node": "in3",
        },
        "to": Object {
          "block": "B",
          "node": "in2",
        },
      },
    },
  ],
}
`;

exports[`Logic Composition verify links when block delete 3`] = `
Map {
  "out1" => Array [
    Object {
      "block": Object {
        "_id": "C",
        "logicKey": "_std_ADDSTRING",
        "xy": Array [
          0,
          0,
        ],
      },
      "node": "out",
      "path": Object {
        "_id": "p6",
        "from": Object {
          "block": "C",
          "node": "out",
        },
        "to": Object {
          "node": "out1",
        },
      },
    },
  ],
}
`;

exports[`Logic Composition verify links when path delete 1`] = `
Object {
  "_id": "ADDSTRING_COMPOUND",
  "blocks": Array [
    Object {
      "_id": "A",
      "logicKey": "ADDSTRING",
      "xy": Array [
        0,
        0,
      ],
    },
    Object {
      "_id": "B",
      "logicKey": "ADDSTRING_NESTED",
      "xy": Array [
        0,
        0,
      ],
    },
    Object {
      "_id": "C",
      "logicKey": "_std_ADDSTRING",
      "xy": Array [
        0,
        0,
      ],
    },
  ],
  "paths": Array [],
}
`;

exports[`Logic Composition verify links when path delete 2`] = `
Map {
  "in1" => Array [],
  "in2" => Array [],
  "in3" => Array [],
}
`;

exports[`Logic Composition verify links when path delete 3`] = `
Map {
  "out1" => Array [],
}
`;

exports[`Logic Composition verify links when path delete 4`] = `
Map {
  "in1" => Array [],
  "in2" => Array [],
}
`;

exports[`Logic Composition verify links when path delete 5`] = `
Map {
  "out1" => Array [],
}
`;

exports[`Logic Composition verify path initialization in blocks 1`] = `
Map {
  "in1" => Array [
    Object {
      "block": Object {
        "_id": "A",
        "logicKey": "ADDSTRING",
        "xy": Array [
          0,
          0,
        ],
      },
      "node": "in1",
      "path": Object {
        "_id": "p0",
        "from": Object {
          "node": "in1",
        },
        "to": Object {
          "block": "A",
          "node": "in1",
        },
      },
    },
  ],
  "in2" => Array [
    Object {
      "block": Object {
        "_id": "A",
        "logicKey": "ADDSTRING",
        "xy": Array [
          0,
          0,
        ],
      },
      "node": "in2",
      "path": Object {
        "_id": "p1",
        "from": Object {
          "node": "in2",
        },
        "to": Object {
          "block": "A",
          "node": "in2",
        },
      },
    },
    Object {
      "block": Object {
        "_id": "B",
        "logicKey": "ADDSTRING_NESTED",
        "xy": Array [
          0,
          0,
        ],
      },
      "node": "in1",
      "path": Object {
        "_id": "p2",
        "from": Object {
          "node": "in2",
        },
        "to": Object {
          "block": "B",
          "node": "in1",
        },
      },
    },
  ],
  "in3" => Array [
    Object {
      "block": Object {
        "_id": "B",
        "logicKey": "ADDSTRING_NESTED",
        "xy": Array [
          0,
          0,
        ],
      },
      "node": "in2",
      "path": Object {
        "_id": "p3",
        "from": Object {
          "node": "in3",
        },
        "to": Object {
          "block": "B",
          "node": "in2",
        },
      },
    },
  ],
}
`;

exports[`Logic Composition verify path initialization in blocks 2`] = `
Map {
  "out1" => Array [
    Object {
      "block": Object {
        "_id": "C",
        "logicKey": "_std_ADDSTRING",
        "xy": Array [
          0,
          0,
        ],
      },
      "node": "out",
      "path": Object {
        "_id": "p6",
        "from": Object {
          "block": "C",
          "node": "out",
        },
        "to": Object {
          "node": "out1",
        },
      },
    },
  ],
}
`;

exports[`Logic Composition verify path initialization in blocks 3`] = `
Map {
  "in1" => Array [
    Object {
      "block": undefined,
      "node": "in1",
      "path": Object {
        "_id": "p0",
        "from": Object {
          "node": "in1",
        },
        "to": Object {
          "block": "A",
          "node": "in1",
        },
      },
    },
  ],
  "in2" => Array [
    Object {
      "block": undefined,
      "node": "in2",
      "path": Object {
        "_id": "p1",
        "from": Object {
          "node": "in2",
        },
        "to": Object {
          "block": "A",
          "node": "in2",
        },
      },
    },
  ],
}
`;

exports[`Logic Composition verify path initialization in blocks 4`] = `
Map {
  "out1" => Array [
    Object {
      "block": Object {
        "_id": "C",
        "logicKey": "_std_ADDSTRING",
        "xy": Array [
          0,
          0,
        ],
      },
      "node": "in1",
      "path": Object {
        "_id": "p4",
        "from": Object {
          "block": "A",
          "node": "out1",
        },
        "to": Object {
          "block": "C",
          "node": "in1",
        },
      },
    },
  ],
}
`;
